Add CRI architecture to cluster concepts (#30606)
kubernetes/website: content/en/docs/concepts/architecture/cri.md

---
title: Container Runtime Interface (CRI)
content_type: concept

> weight: 10

Reviewer: I'd make the weight higher (de-emphasizing the page).

Author: Changed it to `50`.

---

<!-- overview -->

> 

Reviewer: ```suggestion
You need a working
{{<glossary_tooltip text="container runtime" term_id="container-runtime">}} on
each Node in your cluster, so that the
{{< glossary_tooltip text="kubelet" term_id="kubelet" >}} can launch
{{< glossary_tooltip text="Pods" term_id="pod" >}} and their containers.

```
(if accepting this, move `{{< glossary_tooltip text="kubelet" term_id="kubelet" >}}` and `{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}` here)

> The Kubernetes Container Runtime Interface (CRI) defines the main

Reviewer: Better: add this or some similar text to the glossary, and then reuse that here with a `glossary_definition` shortcode.

Or use the same text twice (not shortcode).

nit: is the CRI part of Kubernetes/

Author: Note to myself: I'll do this next week, I'm OOO till Monday

[gRPC](https://grpc.io) protocol for the communication between the
[cluster components](/docs/concepts/overview/components/#node-components)
{{< glossary_tooltip text="kubelet" term_id="kubelet" >}} and
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}.

The CRI is a plugin interface which enables the kubelet to use a wide variety of
container runtimes, without having a need to recompile the cluster components.

<!-- body -->

> ## The API

Reviewer: (nit)
```suggestion
## The API {#api}
```

{{< feature-state for_k8s_version="v1.23" state="stable" >}}

> The latest version of the CRI is `v1`, which is being used by the kubelet as
> default expected version container runtimes have to fulfill. The kubelet acts as
> a client when connecting to the container runtime via gRPC. This means, if a
> container runtime does not support `v1`, then the kubelet tries to connect by
> using the previous version `v1alpha2`. If this connection fails as well, then
> the kubelet gives up. End users have therefore not to take action because the
> kubelet will automatically choose the correct CRI version to select.

Reviewer: ```suggestion
The kubelet acts as a client when connecting to the container runtime via gRPC.
For Kubernetes v{{< skew currentVersion >}}, the kubelet prefers to use CRI `v1`.
If a container runtime does not support `v1` of the CRI, then the kubelet tries to
negotiate any older supported version.
The v{{< skew currentVersion >}} kubelet can also negotiate CRI `v1alpha2`.
If the kubelet cannot negotiate a supported CRI version, the kubelet gives up
and doesn't register as a node.
```

I prefer this because it makes it more clear what support we're describing, and if
you end up looking at docs for a future Kubernetes release that doesn't support
`v1alpha2` it will be more obvious why not (we'd change this page for that release, of course).

If “doesn't register as a node” is wrong, let's document what the kubelet actually does in that situation
(eg fails to launch Pods, taints itself, whatever).

---
title: Container Runtime Interface (CRI)
content_type: concept
weight: 50
---

<!-- overview -->

You need a working
{{<glossary_tooltip text="container runtime" term_id="container-runtime">}} on
each Node in your cluster, so that the
{{< glossary_tooltip text="kubelet" term_id="kubelet" >}} can launch
{{< glossary_tooltip text="Pods" term_id="pod" >}} and their containers.

{{< glossary_definition term_id="container-runtime-interface" length="all" >}}

> The CRI is a plugin interface which enables the kubelet to use a wide variety of

Reviewer 2: perhaps make this sentence first and one about the need for working container runtime second?

Author: Switched them as suggested :+1:

container runtimes, without having a need to recompile the cluster components.

<!-- body -->

## The API {#api}

{{< feature-state for_k8s_version="v1.23" state="stable" >}}

> The kubelet acts as a client when connecting to the container runtime via gRPC.

Reviewer 2: maybe mention that image services may be a separate endppiont and link to the kubelet config setting for it

Author: I added a note about both flags and linked to the kubelet docs.

For Kubernetes v{{< skew currentVersion >}}, the kubelet prefers to use CRI `v1`.
If a container runtime does not support `v1` of the CRI, then the kubelet tries to
negotiate any older supported version.

> The v{{< skew currentVersion >}} kubelet can also negotiate CRI `v1alpha2`.

Reviewer 2: Maybe mention that this is the behavior for now and is subject for deprecation.

Reviewer: We typically [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) unless we're committing to them.

Reviewer 2: ok, it is already printing that v1alpha2 is deprecated. So worth mentioning it here.

This is what printed on fallback:

```
remote_runtime.go:118] "Falling back to CRI v1alpha2 runtime API (deprecated)"
remote_image.go:94] "Falling back to CRI v1alpha2 image API (deprecated)"
```

Reviewer 2: twice because of image services (other comment)

Author: Added a note that it's now deprecated :+1:
